cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -128,10 +126,63 @@ public Collection<StructuredRow> multiRead(Collection<? extends Collection<Field
   @Override
   public CloseableIterator<StructuredRow> scan(Range keyRange, int limit) throws InvalidFieldException {
     LOG.trace("Table {}: Scan range {} with limit {}", schema.getTableId(), keyRange, limit);
-    return new LimitIterator(Collections.singleton(new ScannerIterator(getScanner(keyRange), schema)).iterator(),
-                             limit);
+    return new LimitIterator(Collections.singleton(getFilterByRangeIterator(keyRange)).iterator(), limit);
   }
 
+  /**
+   * Scan the nosql db with a key range. It gets the longest prefix keys,
+   * scan the DB and do a post filtering on the result.
+   *
+   * @param keyRange key range to scan
+   * @return the iterator after filtering
+   * @throws InvalidFieldException if the key is not primary key
+   */
+  private FilterByRangeIterator getFilterByRangeIterator(Range keyRange) {
+    Range prefixRange = getLongestPrefixRange(keyRange);
+    ScannerIterator scannerIterator = new ScannerIterator(getScanner(prefixRange), schema);
+
+    Range filterRange = getFilterRange(keyRange, prefixRange);
+    return new FilterByRangeIterator(scannerIterator, Collections.singleton(filterRange));
```

Comment:
It would be pretty common for filterRange to be empty (when input range is all primary key). Could you skip creating FilterByRangeIterator and returning ScannerIterator in this case?

Reply:
updated